kaspresso/src/main/kotlin/com/kaspersky/kaspresso/autoscroll/AutoScrollFallbackProviderImpl.kt

```diff
@@ -0,0 +1,129 @@
+     *
+     * @param interaction the instance of [ViewInteraction] interface to perform actions and assertions.
+     * @param action the actual action on the interacted view.
+     *
+     * @throws Throwable if the exception caught while invoking [action] is not allowed via [params].
+     * @return the result of [action] invocation.
+     */
+    @Throws(Throwable::class)
+    override fun <T> withAutoScroll(interaction: ViewInteraction, action: () -> T): T {
+        return try {
+            action.invoke()
+        } catch (error: Throwable) {
+            if (error.isAllowed(params.allowedExceptions)) {
+                return scroll(interaction, action, error)
+            }
+            throw error
+        }
+    }
+
+    /**
+     * Performs the autoscrolling functionality. Performs scroll and re-invokes the given [action].
+     *
+     * @param interaction the instance of [ViewInteraction] interface to perform actions and assertions.
+     * @param action the actual action on the interacted view.
+     * @param cachedError the error to be thrown if autoscroll would not help.
+     *
+     * @throws cachedError if autoscroll action did not help.
+     * @return the result of [action] invocation.
+     */
+    @Throws(Throwable::class)
+    override fun <T> scroll(interaction: ViewInteraction, action: () -> T, cachedError: Throwable): T {
+        return try {
+            interaction.perform(FallbackScrollToAction())
+            logger.i("View autoScroll fallback successfully performed.")
+            action.invoke()
+        } catch (error: Throwable) {
+            throw cachedError
+        }
+    }
+}
+
+class FallbackScrollToAction : ViewAction {
+    override fun getConstraints(): Matcher<View> {
+        return Matchers.allOf(
+            ViewMatchers.withEffectiveVisibility(ViewMatchers.Visibility.VISIBLE),
+            ViewMatchers.isDescendantOfA(
+                Matchers.anyOf(
+                    ViewMatchers.isAssignableFrom(NestedScrollView::class.java),
+                    ViewMatchers.isAssignableFrom(ScrollView::class.java),
+                    ViewMatchers.isAssignableFrom(HorizontalScrollView::class.java),
+                    ViewMatchers.isAssignableFrom(ListView::class.java)
+                )
+            )
+        )
+    }
+
+    private tailrec fun findFirstParentScrollView(view: View?): View? {
+        return if (view is ScrollView || view is NestedScrollView || view is ListView) {
+            view
+        } else if (view?.parent != null) {
+            findFirstParentScrollView(view.parent as View)
+        } else {
+            null
+        }
+    }
+
+    override fun perform(uiController: UiController, view: View) {
+        if (isDisplayingAtLeast(100).matches(view)) {
+            Log.i(TAG, "View is already displayed. Returning.")
+            return
+        }
+
+        val scrollView = findFirstParentScrollView(view)
+        scrollView?.scrollTo(0, view.top - scrollView.paddingTop)
```

Comment:
here is the magic. Obviously, for horizontalScrollViews we will need paddingEnd instead